consensus/event_loop.go

```diff
@@ -0,0 +1,854 @@
+	// to do here.
+	if fc.status != eventLoopStatusRunning {
+		return
+	}
+
+	// Check for race conditions where the view advanced at the exact moment this task began
+	// or we have already signaled for this view. If so, then there's nothing to do here.
+	if fc.currentView != blockConstructionView {
+		return
+	}
+
+	// Mark that the crank timer has elapsed
+	fc.hasCrankTimerRunForCurrentView = true
+
+	// Check if the conditions are met to construct a QC from votes for the chain tip. If so,
+	// we send a signal to the server and cancel the crank timer task. The server will
+	// reschedule the task when it advances the view.
+	if voteQCEvent := fc.tryConstructVoteQCInCurrentView(); voteQCEvent != nil {
+		// Signal the server that we can construct a QC for the chain tip
+		fc.hasConstructedQCInCurrentView = true
+		fc.Events <- voteQCEvent
+		return
+	}
+
+	// Check if we have enough timeouts to build an aggregate QC for the previous view. If so,
+	// we send a signal to the server and cancel all scheduled tasks.
+	if timeoutQCEvent := fc.tryConstructTimeoutQCInCurrentView(); timeoutQCEvent != nil {
+		// Signal the server that we can construct a timeout QC for the current view
+		fc.hasConstructedQCInCurrentView = true
+		fc.Events <- timeoutQCEvent
+		return
+	}
+}
+
+// tryConstructVoteQCInCurrentView is a helper function that attempts to construct a QC for the tip block
+// so that it can be proposed in a block in the current view. The function internally performs all view and vote
+// validations to ensure that the resulting QC is valid. If a QC can be constructed, the function returns
+// the signers list and aggregate signature that can be used to construct the QC.
+//
+// This function must be called while holding the event loop's lock.
+func (fc *FastHotStuffEventLoop) tryConstructVoteQCInCurrentView() *FastHotStuffEvent {
+	// If currentView != tipBlock.View + 1, then we have timed out at some point, and can no longer
+	// construct a block with a QC of votes for the tip block.
+	tipBlock := fc.tip.block
+	if fc.currentView != tipBlock.GetView()+1 {
+		return nil
+	}
+
+	// Fetch the validator set at the tip.
+	validatorSet := fc.tip.validatorSet
+
+	// Compute the chain tip's signature payload.
+	voteSignaturePayload := GetVoteSignaturePayload(tipBlock.GetView(), tipBlock.GetBlockHash())
+
+	// Fetch the validator votes for the tip block.
+	votesByValidator := fc.votesSeen[voteSignaturePayload]
+
+	// Compute the total stake and total stake with votes
+	totalStake := uint256.NewInt()
+	totalVotingStake := uint256.NewInt()
+
+	// Track the signatures and signers list for the chain tip
+	signersList := bitset.NewBitset()
+	signatures := []*bls.Signature{}
+
+	// Iterate through the entire validator set and check if each one has voted for the tip block. Track
+	// all voters and their stakes.
+	for ii, validator := range validatorSet {
+		totalStake = uint256.NewInt().Add(totalStake, validator.GetStakeAmount())
```

Comment:
Is this a bug? It looks like you add a validator's stake to the total even if they haven't voted for this block. I think you want to move this line to after you've checked the vote?

Reply:
Not a bug.

We dynamically sum the entire validator set's stake here. The number doesn't really change so I guess we could compute it once and cache it, but since we iterate over the entire validator set here, we may as well sum it here.

Remember, from the perspective of the FastHotStuffEventLoop, every block height can its own unique validator set. So there's no single "GlobalStake" value. In practice, the validator set will only change for block heights on epoch transitions, but the FastHotStuffEventLoop shouldn't know when that is.

Reply:
Sorry that was a dumb question. I think I was confusing totalStake for totalVotingStake...